the-collab-lab/tcl-6-smart-shopping-list
# sk-es-alert-user

_For an example of how to fill this template out, [see this Pull Request](https://github.com/the-collab-lab/tcl-3-smart-shopping-list/pull/44)._ 
## To Do
- [x] work with group on how to store data (collections and docs) in firebase
- [x] Clean out test data
- [x] get name field(s) from database
- [x] get the name from the form input field
- [x] compare name (shopping list item) from database with name user entered
- [x] normalize the 2 strings that are being compared
- [x] Use .toLower() - make all lowercase, .trim() - remove beginning and end spaces, and .replace('NEED regular expression', '') - replace punctuation and special characters with nothing
- [x] if input name doesn't match save a new item to the shopping list
- [x] else show the user an error message

## Description

This PR adds new functionality to the `AddItemForm` component. The `AddItemForm` now has the ability to display an error message when a user inputs an item that has already been saved to the database. 

### Steps to accomplish this feature:

We've updated the state of `AddItemForm` to hold a `name_normalize` property. This was done to differentiate between the user input: `name` from `name_normalize` when saved to the database. The name_normalize is a feature that accounts for any whitespace, capitalization, and special characters using a regex pattern. The regex pattern is used on the incoming item added by the user.

Once an item is added, it's been submitted to the database where it can be retrieved to be filtered. We are able to do this by the use of the `FirestoreCollection` tag within our form. Firestore collection is able to access the `items` collection and create a simple query to filter the normalized name to the name saved to the database. The filtering will generate an error message with: "`item name` has already been added to the shopping list" if the input is already in the database. 





## Related Issue

closes #8 

## Acceptance Criteria

* Show an error message if the user tries to submit a new item that has the exact same name as an existing item
* Show an error message if the user tries to submit a new item that has the same name as an existing item, where capitalization has been normalized and punctuation has been removed
* The user’s original input should be what gets saved in the database

## Type of Changes



|     | Type                       |
| --- | -------------------------- |
|    | :bug: Bug fix              |
|  ✓ | :sparkles: New feature     |
|    | :hammer: Refactoring       |
|    | :100: Add tests            |
|    | :link: Update dependencies |
|    | :scroll: Docs              |

## Updates

### Before
<img width="626" alt="Screen Shot 2020-04-23 at 6 22 10 PM" src="https://user-images.githubusercontent.com/48725513/80165089-7ea85500-858f-11ea-851b-62948fa7e319.png">




### After
<img width="638" alt="Screen Shot 2020-04-23 at 6 17 24 PM" src="https://user-images.githubusercontent.com/48725513/80164949-240ef900-858f-11ea-85ad-aff2bcb3817c.png">




## Testing Steps / QA Criteria

1. Git pull the branch using `git pull origin sk-es-alert-user` and `git checkout sk-es-alert-user` to switch to the branch.
2. Run npm install to add dependencies, then npm start to run the app.
3. Type an item into the name input field
4. Select a scheduled time and hit submit.
5. To test If It's working properly: Select the same item again, you can even add a special character or capital letter. 
6. An error message should appear. 

**Note:** `apple`, `Apple1`, and `Peanut Butter` are currently saved in the database and will present an error message if typed into the name field.
